src/commands/utils/graph.ts

```diff
@@ -0,0 +1,230 @@
+import { AuthenticationProvider, Client } from "@microsoft/microsoft-graph-client";
+import { Errorable, getErrorMessage } from "./errorable";
+import { getDefaultScope, getEnvironment } from "../../auth/azureAuth";
+import { GetAuthSessionOptions, ReadyAzureSessionProvider } from "../../auth/types";
+
+const federatedIdentityCredentialIssuer = "https://token.actions.githubusercontent.com";
+const federatedIdentityCredentialAudience = "api://AzureADTokenExchange";
+
+type GraphListResult<T> = {
+    value: T[];
+};
+
+export type ApplicationParams = {
+    displayName: string;
+};
+
+export type Application = ApplicationParams & {
+    appId: string;
+    id: string;
+};
+
+export type ServicePrincipalParams = {
+    appId: string;
+    displayName?: string;
+};
+
+export type ServicePrincipal = ServicePrincipalParams & {
+    id: string;
+    displayName: string;
+};
+
+export type FederatedIdentityCredentialParams = {
+    name: string;
+    subject: string;
+    issuer: string;
+    description: string;
+    audiences: string[];
+};
+
+export type FederatedIdentityCredential = FederatedIdentityCredentialParams & {
+    id: string;
+};
+
+export function createGraphClient(sessionProvider: ReadyAzureSessionProvider): Client {
+    // The "Visual Studio Code" application id.
+    const applicationClientId = "aebc6443-996d-45c2-90f0-388ff96faa56";
+
+    const baseUrl = getMicrosoftGraphClientBaseUrl();
+    const authProvider: AuthenticationProvider = {
+        getAccessToken: async (options) => {
+            const authSessionOptions: GetAuthSessionOptions = {
+                scopes: options?.scopes || [getDefaultScope(baseUrl)],
+                applicationClientId,
+            };
+
+            const session = await sessionProvider.getAuthSession(authSessionOptions);
+            return session.succeeded ? session.result.accessToken : "";
+        },
+    };
+
+    return Client.initWithMiddleware({ baseUrl, authProvider });
+}
+
+function getMicrosoftGraphClientBaseUrl(): string {
+    const environment = getEnvironment();
+    // Environments are from here: https://github.com/Azure/ms-rest-azure-env/blob/6fa17ce7f36741af6ce64461735e6c7c0125f0ed/lib/azureEnvironment.ts#L266-L346
+    // They do not contain the MS Graph endpoints, whose values are here:
+    // https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/d365ab1d68f90f2c38c67a5a7c7fe54acfc2584e/src/Constants.ts#L28
+    switch (environment.name) {
+        case "AzureChinaCloud":
+            return "https://microsoftgraph.chinacloudapi.cn";
+        case "AzureUSGovernment":
+            return "https://graph.microsoft.us";
+        case "AzureGermanCloud":
+            return "https://graph.microsoft.de";
+    }
+
+    return "https://graph.microsoft.com";
+} //TODO: Proper error handling
```

Comment:
💡 = ) What is the proper `error handling` here as mentioned in comment.

Reply:
Went ahead and removed that. Had added that initially thinking all environments were not accounted for. Thanks!